Store new dav chunks directly, no hooks

PVince81 opened the pull request:
## Description
Store new dav chunks directly instead of going through File::put() which
would run all hooks. Hooks are now only triggered at the end of the
transfer when the future file is moved to the final file, in which case
the assemble data goes through File::put() and triggers the hooks.

**Note:** this likely doesn't solve the problem for storage wrappers. For storage wrappers, these need to verify whether the target path is outside of "files" and bypass operation in such cases. Not part of this PR, needs to be implemented in every storage wrapper.

## Related Issue
https://github.com/owncloud/enterprise/issues/2229

## Motivation and Context
Prevent hooks to be fired for every chunk.

## How Has This Been Tested?
Upload 100 MB file in web UI.

Before this fix: oc_activity polluted with extra entries for each chunk
After this fix: oc_activity only

Also works with encryption.

Can someone test this with object store ? @DeepDiver1975 @SergioBertolinSG 

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
Store new dav chunks directly, no hooks

Store new dav chunks directly instead of going through File::put() which
would run all hooks. Hooks are now only triggered at the end of the
transfer when the future file is moved to the final file, in which case
the assemble data goes through File::put() and triggers the hooks.

PVince81 commented:
I've set this to p1 because triggering hooks for every chunk is likely to cause unwanted and undiscovered side effects.

DeepDiver1975 commented:
checksums are not properly computed now?

![bildschirmfoto von 2017-08-25 09-07-41](https://user-images.githubusercontent.com/1005065/29703213-e700ed84-8974-11e7-9068-9e79f3519fdb.png)

PVince81 commented:
@DeepDiver1975 I need to look into this. Some tests were expecting checksums for every chunks and need to be adjusted.

PVince81 commented:
apparently there isn't even a test for new dav final move with checksum, I'll add one

Commit:
Adjust checksum tests for new dav chunking

Don't send checksums for each chunk but for the final move.
Added test for the final move.
Removed test code for chunk checksum.

PVince81 commented:
@DeepDiver1975 tests adjusted, see https://github.com/owncloud/core/pull/28795/commits/f09c06ccccb962bf4caa5af10767db696da22027

PVince81 commented:
@owncloud/qa can you guys test this with objectstore ? I expect uploaded chunks to be stored in object store by default and should still work .

PVince81 commented:
> @PVince81 Who is finle? :)

who are you talking about ? never heard of such person 🕶

jvillafanez commented on apps/dav/lib/Connector/Sabre/Node.php:
```diff
@@ -394,4 +394,8 @@ protected function sanitizeMtime ($mtimeFromRequest) {
 		return $mtime;
 	}
 
+	public function getView() {
```
I'm not sure if we should expose this. Taking into account that we only need one method of the view, we can wrap that method we need inside a specific method in the node. This way we don't need to expose the underlying view object.

PVince81 replied:
you mean creating a method `putDirectly($path, $data)` and call that from outside ?

jvillafanez replied:
yes, that's what I mean,

Commit:
Directory::createFileDirectly

PVince81 commented:
@jvillafanez https://github.com/owncloud/core/pull/28795/commits/44e72ac9b149d67d5c0c402049cfc676c3823350

PVince81 commented:
stable10: https://github.com/owncloud/core/pull/28817

phil-davis commented on apps/dav/lib/Upload/UploadFolder.php:
```diff
@@ -34,8 +34,8 @@ function __construct(Directory $node) {
-		// TODO: verify name - should be a simple number
-		$this->node->createFile($name, $data);
+		// need to bypass hooks for individual chunks
+		$this->node->createFileDirectly($name, $data);
```
@PVince81 @individual-it I am having trouble getting firewall file upload tests to pass now. ``git bisect`` is kindly telling me that the commit(s) in this PR are the start of my troubles.
I run the following firewall test:
```
bash tests/travis/start_behat_tests.sh -c apps/firewall/tests/ui/config/behat.yml --feature apps/firewall/tests/ui/features/limituploadbasedonsize.feature:23
```
It fails with:
```
Then it should not be possible to upload a file with the size of "40000" bytes using "any" dav path and any chunking             # FileFirewallContext::itShouldNotBePossibleToUploadAFileWithTheSizeOf()
      using v2 dav path and v2 chunking
      Failed asserting that 500 matches expected 403.
```
owncloud.log has:
```
{"reqId":"ROzXXzvP7zhMDM7qa0yn","level":4,"time":"2017-08-29T17:26:27+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"webdav","method":"PUT","url":"\/remote.php\/dav\/uploads\/admin\/chunking-2117\/0","message":"Exception: {\"Message\":\"Access to this resource has been forbidden by a file firewall rule.\",\"Exception\":\"OCP\\\\Files\\\\ForbiddenException\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/phil\\\/core\\\/apps\\\/firewall\\\/lib\\\/storagewrapper.php(68): OCA\\\\Firewall\\\\Firewall->checkRulesForFiles(Array)\\n#1 \\\/home\\\/phil\\\/core\\\/apps\\\/firewall\\\/lib\\\/storagewrapper.php(334): OCA\\\\Firewall\\\\StorageWrapper->checkFirewall('uploads')\\n#2 \\\/home\\\/phil\\\/core\\\/lib\\\/private\\\/Files\\\/View.php(648): OCA\\\\Firewall\\\\StorageWrapper->fopen('uploads\\\/chunkin...', 'w')\\n#3 \\\/home\\\/phil\\\/core\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php(472): OC\\\\Files\\\\View->file_put_contents('\\\/chunking-2117\\\/...', Resource id #44)\\n#4 \\\/home\\\/phil\\\/core\\\/apps\\\/dav\\\/lib\\\/Upload\\\/UploadFolder.php(38): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->createFileDirectly('0', Resource id #44)\\n#5 \\\/home\\\/phil\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(1095): OCA\\\\DAV\\\\Upload\\\\UploadFolder->createFile('0', Resource id #44)\\n#6 \\\/home\\\/phil\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(525): Sabre\\\\DAV\\\\Server->createFile('uploads\\\/admin\\\/c...', Resource id #44, NULL)\\n#7 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpPut(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#8 \\\/home\\\/phil\\\/core\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#9 \\\/home\\\/phil\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:PUT', Array)\\n#10 \\\/home\\\/phil\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#11 \\\/home\\\/phil\\\/core\\\/apps\\\/dav\\\/lib\\\/Server.php(236): Sabre\\\\DAV\\\\Server->exec()\\n#12 \\\/home\\\/phil\\\/core\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#13 \\\/home\\\/phil\\\/core\\\/remote.php(175): require_once('\\\/home\\\/phil\\\/core...')\\n#14 {main}\",\"File\":\"\\\/home\\\/phil\\\/core\\\/apps\\\/firewall\\\/lib\\\/firewall.php\",\"Line\":192,\"User\":\"admin\"}"}
```
and other "forbidden" entries before it that mention ``HTTP\\\/1.1 403`` before ``Access to this resource has been forbidden by a file firewall rule.``

If I revert this line, so it just does:
```
$this->node->createFile($name, $data);
```
then the test passes (it gets a 403 instead of a 500), and in owncloud.log is:
```
{"reqId":"B2ujE6POVaStc2kjdRhe","level":4,"time":"2017-08-29T17:23:48+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"webdav","method":"PUT","url":"\/remote.php\/dav\/uploads\/admin\/chunking-3337\/0","message":"Exception: {\"Message\":\"HTTP\\\/1.1 403 Access to this resource has been forbidden by a file firewall rule.\",\"Exception\":\"OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Exception\\\\Forbidden\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/phil\\\/core\\\/apps\\\/dav\\\/lib\\\/Upload\\\/UploadFolder.php(38): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->createFile('0', Resource id #43)\\n#1 \\\/home\\\/phil\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(1095): OCA\\\\DAV\\\\Upload\\\\UploadFolder->createFile('0', Resource id #43)\\n#2 \\\/home\\\/phil\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(525): Sabre\\\\DAV\\\\Server->createFile('uploads\\\/admin\\\/c...', Resource id #43, NULL)\\n#3 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpPut(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/home\\\/phil\\\/core\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#5 \\\/home\\\/phil\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:PUT', Array)\\n#6 \\\/home\\\/phil\\\/core\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#7 \\\/home\\\/phil\\\/core\\\/apps\\\/dav\\\/lib\\\/Server.php(236): Sabre\\\\DAV\\\\Server->exec()\\n#8 \\\/home\\\/phil\\\/core\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#9 \\\/home\\\/phil\\\/core\\\/remote.php(175): require_once('\\\/home\\\/phil\\\/core...')\\n#10 {main}\",\"File\":\"\\\/home\\\/phil\\\/core\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php\",\"Line\":168,\"User\":\"admin\"}"} 
```
which now has the ``HTTP\\\/1.1 403`` in the start of the exception message.

There is some interaction here between what this PR is doing, and how the firewall, Sabre... are getting triggered and what code paths end up putting in the ``403`` HTTP status code number.

Ref also https://github.com/owncloud/firewall/pull/397

Needs sorting out.

PVince81 replied:
If the firewall is using hooks then those aren't triggered any more.
